src/pip/_internal/cli/req_command.py

```diff
@@ -232,15 +228,14 @@ def determine_resolver_variant(options):
     @classmethod
     def make_requirement_preparer(
         cls,
-        temp_build_dir,  # type: TempDirectory
-        options,  # type: Values
-        req_tracker,  # type: RequirementTracker
-        session,  # type: PipSession
-        finder,  # type: PackageFinder
-        use_user_site,  # type: bool
-        download_dir=None,  # type: str
-    ):
-        # type: (...) -> RequirementPreparer
+        temp_build_dir: TempDirectory,
+        options: Values,
+        req_tracker: RequirementTracker,
+        session: PipSession,
+        finder: PackageFinder,
+        use_user_site: bool,
+        download_dir: str = None,
```

Comment:
Hmm, why is Mypy OK with this? This should be `download_dir: Optional[str]`.

Reply:
I thought the same. Maybe we can try to annotate `download_dir` just like you said, and see how Mypy reacts.

Reply:
I will try to apply your suggestions, with the hope that Mypy won't fail now.

Reply:
An early version of the typing PEPs allowed type checkers to infer Optional if there was an initialiser of None (see [here](https://discuss.python.org/t/pep-661-sentinel-values/9126/39)). Maybe mypy is using the old rules (or we're using an old version of mypy?)

Reply:
Might be. Anyway, we must take a look of what is going on with Mypy.

Reply:
Actually, see [this](https://stackoverflow.com/a/62732458/34438)

Reply:
So, omitting `Optional` must not be recommended, right? Interesting.

Reply:
What version of Mypy is testing the PRs? Because using an old one could be harmful, even in other cases.

Reply:
[We’re using 0.800](https://github.com/pypa/pip/blob/7c3abccd22c4bbd6c1da2ae41845ffbd26c1bb17/.pre-commit-config.yaml#L59), so that’s weird.

Reply:
Maybe opening an issue could help.

Reply:
Oh actually, rereading Mypy’s documentation, automatic Optional inference is still the default behaviour (but deprecated), and we need to explicitly set [`--no-implicit-optional`](https://mypy.readthedocs.io/en/stable/command_line.html#cmdoption-mypy-no-implicit-optional) to get the PEP 484 compliant behaviour. Could you try whether setting that would work?

Reply:
> Could you try whether setting that would work?

Do you want me to try the option `--no-implicit-optional` on my own computer, or changing the Mypy options for all GitHub (I mean, `ci.yml`)? I didn't understand that. **Anyway, the idea looks great**.

Reply:
Just try it locally first. If that makes a difference we'll probably need to fix a bunch of more things so that should be done in another PR.

Reply:
Ok. I will try `$ mypy src/pip/_internal/cli` (using my local clone). Is that ok?

Reply:
How critical is it that we typecheck this tightly anyway? Regardless of whether we use `Optional[str]` or `str`, mypy will currently infer `Optional[str]` as the type and do the right thing. At the point where we upgrade `.pre-commit-config.yaml` to use a version of mypy that doesn't infer the `Optional`, we'll get error reports for all of the places affected by the stricter typing, and we can fix them all as part of the move to a newer mypy.

(Disclaimer: I actually find `download_dir: str = None` more readable, and I think the automatic promotion is reasonable, so I'm not unbiased here...)

Reply:
Well... we are just going forward to (maybe) fix this and avoid a future reincidence (including a new discussion). That's why @uranusjr and I are trying to find how to make the Mypy test fit to the PEP 661.

By-the-by, how do I make Git to test a specific branch? (sorry for my ignorance)

Reply:
Yeah, if Mypy is going to make the strict behaviour the default, we’ll need to do this eventually, so this is as good a time as any. This is another reason it should be a separate PR; we can take our time on this.

Reply:
> how do I make Git to test a specific branch?

I always just `git checkout` the branch and run things on it.

Reply:
Oh. Thanks. I thought there was a specific way to do it.

Reply:
> This is another reason it should be a separate PR; we can take our time on this.

I will open an issue at pypa/pip, to discuss on a larger way. What do you think about it?

Reply:
> Ok. I will try `$ mypy src/pip/_internal/cli` (using my local clone)

Finally done. When using the `--no-implicit-optional` flag, it returned the expected output \(but it doesn't fails, because I have changed the `Optional[str]` annotation\)

Reply:
I created #10076 for a larger discussion.